### ISSUE TITLE
String patterns were incorrectly translated

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/molecule-polarity#143]
molecule-polarity-strings_da.json has a problem with pattern strings.  The placeholder names have been changed.

This English string:

```
  "pattern.dipoleDirection": {
    "value": "{{from}} → {{to}}"
  },
```

... is used to fill in these radio button labels, which appear in Preferences:

<img width="108" alt="screenshot_1846" src="https://user-images.githubusercontent.com/3046552/187577224-2f73fd08-3212-4f16-8285-5add746d73ca.png">

The current 'da' translation is:

```
 "pattern.dipoleDirection": {
    "value": "{{fra}} → {{til}}",
```

This makes it impossible to fill in the pattern correctly, and it shows up in the sim UI like this:

<img width="126" alt="screenshot_1845" src="https://user-images.githubusercontent.com/3046552/187576920-0d5462fe-6b0c-4de3-bbab-ef220bee68ab.png">

The problem is that the translator has translated the placeholders, which are not supposed to be changed, and that makes it impossible for the sim to fill in the blanks. Rosetta is supposed to prevent this (@jbphet FYI). Is Rosetta working correctly? Do other translations have this problem?  Should the 'da' trusted translator be notified.

Assigning to @kathy-phet to decide on next steps.

**Comment by pixelzoom** [comment]
I found another broken patterns in molecule-polarity-strings_fr.json. They translated the `symbol` placeholder to `symbole`:

```
  "pattern.symbolName": {
    "value": "{{symbole}}({{name}})",
```

Rosetta is supposed to prevent this.  This commit was made on 7/30/17.

**Comment by pixelzoom** [comment]
Searching .json file under bable/molecule-polarity for `"value": "{{`, I see translated placeholders in:

da (1)
fr (1)
ko (2)
vi (1)
zh_CN (2)

So this is a widespread problem.  I have not inspected string files for other sims.

**Comment by liammulh** [comment]
Commenting here so that I get notifications.

**Comment by liammulh** [comment]
I'm going to check to make sure old rosetta matches the body of the curly brace expressions.

**Comment by pixelzoom** [comment]
The files where this is a problem look like older commits. And I don't see this problem in my newer sims (geometric-optics, natural-selection, fourier-making-waves).  So I'm guessing that these errors pre-date the support that was added to Rosetta to prevent modifying placeholders.  (But it would be good to confirm.)

**Comment by pixelzoom** [comment]
@liammulh status?

**Comment by zepumph** [comment]
Discussed today during PhET-iO meeting. We first want to confirm that this will not be possible in the new rosetta (translating the template var names), and then we want to change this since it is effecting the production version of the sim. @kathy-phet asked me to assign @jbphet, but also said that she would create a QA issue to go through and discover all the spots where this is happening in our simulations. I wonder if a script could be better suited to that task? @jbphet do you have something like that (ensure all templated vars in translation babel files match the en name for it)? I can also whip one up if you'd like.

**Comment by jbphet** [comment]
> @jbphet do you have something like that (ensure all templated vars in translation babel files match the en name for it)? I can also whip one up if you'd like.

No, I don't have such a script.  There hasn't been a need before now.

I just checked the validation function on the live translation utility, and it seems to be working.  Screenshot below.  It's possible that the submit button is still working in this situation, but I didn't want to press it and create a bad translation, so I'll look into it later.

I'll do more investigation later.

![image](https://user-images.githubusercontent.com/3082831/207963952-3c6f0249-7e77-44f2-9755-de430c579e0c.png)

**Comment by kathy-phet** [comment]
Rather than ask QA, I'm going to as @oliver-phet to go through. I'll make a new issue though for the fix to the live translations.

**Comment by zepumph** [comment]
Hello! I spent 15 minutes and wrote a script that should every translation in which the current value of a translated string key does not match the english version in master. I think it will at least be a good starting point, but it is worth noting that there may be a disconnect between published sims and master. 

<details>
<summary>the script</summary>

```js
const fs = require( 'fs' );
const _ = require( 'lodash' );
const x = fs.readdirSync( '../babel' );


const templateVarRegex = /{\{?[\w\d]+\}\}?/g;
const openRegex = /{/g;
const closeRegex = /}/g;

function getTemplatedVars( string ) {

  const matches = string.match( templateVarRegex ) || [];

  // Don't match {HI}} for whatever reason that may be
  return matches.filter( templateVar => templateVar.match( openRegex ).length === templateVar.match( closeRegex ).length );
}

// @returns map where string key is key, and list of template vars in it are the value.
function getTemplateVarsInEnglishStringFile( repo ) {

  const englishStringFileName = `../${repo}/${repo}-strings_en.json`;
  const englishStrings = JSON.parse( fs.readFileSync( englishStringFileName ).toString() );

  // Record<stringKey, Array<templateVarString>>
  const keyTemplateMap = {};
  Object.keys( englishStrings ).forEach( key => {
    const stringValue = englishStrings[ key ].value;
    if ( stringValue ) {
      const templateVars = getTemplatedVars( stringValue );
      if ( templateVars.length > 0 ) {
        keyTemplateMap[ key ] = templateVars;
      }
    }
  } );

  return keyTemplateMap;
}

x.forEach( dir => {
  const babelRepoDir = `../babel/${dir}`;
  if ( !dir.startsWith( '_' ) && !dir.startsWith( '.' ) && fs.statSync( babelRepoDir ).isDirectory() ) {


    const repo = dir;
    const templateVarsMap = getTemplateVarsInEnglishStringFile( repo );

    fs.readdirSync( babelRepoDir ).forEach( translatedStringFileContents => {
      const translatedStrings = JSON.parse( fs.readFileSync( `../babel/${dir}/${translatedStringFileContents}` ).toString() );
      Object.keys( templateVarsMap ).forEach( stringKey => {

        if ( translatedStrings[ stringKey ] ) {

          const translatedStringValue = translatedStrings[ stringKey ].value;
          const translatedTemplateVars = getTemplatedVars( translatedStringValue );
          const englishTemplatedVars = templateVarsMap[ stringKey ];

          // Set because order doesn't matter
          if ( !_.isEqual( new Set( englishTemplatedVars ), new Set( translatedTemplateVars ) ) ) {
            console.log( `${translatedStringFileContents}:\n`, 'english vars:', englishTemplatedVars, '\n translated vars:', translatedTemplateVars, '\n' );
          }
        }
      } );
    } );
  }
} );
```
</details>

<details>
<summary>Results</summary>


beers-law-lab-strings_ru.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

beers-law-lab-strings_ru.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

beers-law-lab-strings_ru.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

beers-law-lab-strings_ru.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

charges-and-fields-strings_gu.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

circuit-construction-kit-common-strings_nb.json:
 english vars: [ '{{resistance}}' ] 
 translated vars: [] 

circuit-construction-kit-common-strings_ta.json:
 english vars: [ '{{resistance}}' ] 
 translated vars: [] 

expression-exchange-strings_ko.json:
 english vars: [ '{{number}}' ] 
 translated vars: [] 

expression-exchange-strings_ko.json:
 english vars: [ '{{levelNumber}}' ] 
 translated vars: [] 

expression-exchange-strings_zh_CN.json:
 english vars: [ '{{number}}' ] 
 translated vars: [] 

expression-exchange-strings_zh_CN.json:
 english vars: [ '{{levelNumber}}' ] 
 translated vars: [] 

fluid-pressure-and-flow-strings_gu.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

fluid-pressure-and-flow-strings_gu.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

fluid-pressure-and-flow-strings_gu.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

fluid-pressure-and-flow-strings_gu.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

gene-expression-essentials-strings_es_ES.json:
 english vars: [ '{{geneID}}' ] 
 translated vars: [] 

gene-expression-essentials-strings_fa.json:
 english vars: [ '{{geneID}}' ] 
 translated vars: [] 

gene-expression-essentials-strings_ja.json:
 english vars: [ '{{geneID}}' ] 
 translated vars: [] 

gene-expression-essentials-strings_mk.json:
 english vars: [ '{{geneID}}' ] 
 translated vars: [] 

gene-expression-essentials-strings_ta.json:
 english vars: [ '{{geneID}}' ] 
 translated vars: [] 

joist-strings_nl.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

masses-and-springs-strings_ig.json:
 english vars: [ '{{equalsZero}}' ] 
 translated vars: [] 

masses-and-springs-strings_ig.json:
 english vars: [ '{{gravity}}' ] 
 translated vars: [] 

masses-and-springs-strings_ig.json:
 english vars: [ '{{mass}}' ] 
 translated vars: [] 

molarity-strings_fa.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

molecule-polarity-strings_bs.json:
 english vars: [ '{{from}}', '{{to}}' ] 
 translated vars: [ '{{od}}', '{{do}}' ] 

molecule-polarity-strings_da.json:
 english vars: [ '{{from}}', '{{to}}' ] 
 translated vars: [ '{{fra}}', '{{til}}' ] 

molecule-polarity-strings_fa.json:
 english vars: [ '{{name}}' ] 
 translated vars: [] 

molecule-polarity-strings_ko.json:
 english vars: [ '{{name}}' ] 
 translated vars: [] 

molecule-polarity-strings_ko.json:
 english vars: [ '{{symbol}}', '{{name}}' ] 
 translated vars: [] 

molecule-polarity-strings_ko.json:
 english vars: [ '{{from}}', '{{to}}' ] 
 translated vars: [] 

molecule-polarity-strings_vi.json:
 english vars: [ '{{from}}', '{{to}}' ] 
 translated vars: [] 

molecule-polarity-strings_zh_CN.json:
 english vars: [ '{{name}}' ] 
 translated vars: [] 

molecule-polarity-strings_zh_CN.json:
 english vars: [ '{{symbol}}', '{{name}}' ] 
 translated vars: [] 

molecule-polarity-strings_zh_CN.json:
 english vars: [ '{{from}}', '{{to}}' ] 
 translated vars: [] 

number-play-strings_ht.json:
 english vars: [ '{{language}}' ] 
 translated vars: [ '{{langue}}' ] 

pendulum-lab-strings_nb.json:
 english vars: [ '{{gravity}}' ] 
 translated vars: [ '{{gravitasjon}}' ] 

pendulum-lab-strings_nb.json:
 english vars: [ '{{degrees}}' ] 
 translated vars: [ '{{grader}}' ] 

pendulum-lab-strings_nb.json:
 english vars: [ '{{seconds}}' ] 
 translated vars: [ '{{sekunder}}' ] 

proportion-playground-strings_eu.json:
 english vars: [ '{{price}}' ] 
 translated vars: [ '{{prezioa}}' ] 

proportion-playground-strings_ko.json:
 english vars: [ '{{price}}' ] 
 translated vars: [] 

proportion-playground-strings_vi.json:
 english vars: [ '{{price}}' ] 
 translated vars: [] 

proportion-playground-strings_zh_CN.json:
 english vars: [ '{{price}}' ] 
 translated vars: [] 

scenery-phet-strings_ht.json:
 english vars: [ '{{thing}}' ] 
 translated vars: [ '{{chose}}' ] 

scenery-phet-strings_ht.json:
 english vars: [ '{{thing}}' ] 
 translated vars: [ '{{chose}}' ] 

scenery-phet-strings_ht.json:
 english vars: [ '{{distance}}', '{{units}}' ] 
 translated vars: [ '{{distance}}' ] 

scenery-phet-strings_tg.json:
 english vars: [ '{{distance}}', '{{units}}' ] 
 translated vars: [] 

trig-tour-strings_iw.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

trig-tour-strings_iw.json:
 english vars: [ '{0}', '{1}' ] 
 translated vars: [] 

vegas-strings_ht.json:
 english vars: [ '{0}' ] 
 translated vars: [] 

vegas-strings_mt.json:
 english vars: [ '{0}' ] 
 translated vars: []

**Comment by pixelzoom** [comment]
Not sure why another almost-identical issues was opened (which I transferred to https://github.com/phetsims/babel/issues/18) or why we're positing identical comments in this issues and that one. But this problem is clearly not specific to molecule-polarity, and is better dealt with in babel.  So closing this issue, and we'll track in https://github.com/phetsims/babel/issues/18.

**Comment by pixelzoom** [comment]
Also noting that this does not block https://github.com/phetsims/molecule-polarity/issues/137 (Publish 1.3 with PhET-iO Upgrades).

**Comment by zepumph** [comment]
Yes we should move this one too. It is to confirm that the rosetta gui is set up correctly to not let this issue happen again, and @oliver-phet will take the lead. I'll move it to rosetta.
